### PR TITLE
Rewrote all tournament.Added two types of displaying tournament grid

### PR DESCRIPTION
**Pull request by FoxHanter**

### CODE REVIEW

File: 2018/Pichkur/Championship/Championship/Tour.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+﻿using System.Collections.Generic;
+
+namespace Championship
+{
+    public class Tour
+    {
+        public List<Game> Games = new List<Game>();
+        public List<Team> _extraTeams = new List<Team>();
+        private int _countGames = 1;
+        public Tour(List<Team> teams)
```
Comment (dkblock) [inline 2018/Pichkur/Championship/Championship/Tour.cs]: Split new method by empty line

File: 2018/Pichkur/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.vs
+Properties/
+.config
```
Comment (vikian050194) [inline 2018/Pichkur/.gitignore]: Unignore .config

File: 2018/Pichkur/Championship/Championship/Game.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿namespace Championship
+{
+    public class Game
+    {
+        public Team FirstTeam { get; private set; }
+        public Team SecondTeam { get; private set; }
+        public string Winner { get; private set; }
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/Game.cs]: Change type of Winner

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        {
+            _team = new List<Team>(teams);
+            _team = MakeNameOneLength(_team);
+            _minCursorTop = minCursorTop;
+        }
+
+        public void DrawGrid(TournamentGrid grid, GridType type)
+        {
+            var allTeams = new List<Team>(_team);
+            _rightBorder = 2 * (_nameLength + 3 * (grid.Tours.Count - 1) + 4);
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: Split on parts

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+            for (int i = 0; i < lastTour; i++)
+            {
+                int lastgame = GetMiddleGame(grid.Tours[i].Games.Count);
+
+                for (int j = 0; j < grid.Tours[i].Games.Count; j++)
+                {
+                    var currentGame = grid.Tours[i].Games[j];
+
+                    if (j < lastgame || type == GridType.StandardGrid)
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: Split by grid type

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            {
+                if (names[i].Name.Length < _nameLength)
+                {
+                    names[i].Name = names[i].Name.PadLeft(_nameLength);
+                }
+
+                answer.Add(names[i]);
+            }
+
+            return answer;
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: Rename

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            PrintUpLine(isLeft, i, j, game);
+            PrintMiddleLine(isLeft, i, j, game);
+            PrintDownLine(isLeft, i, j, game);
+        }
+
+        private void PrintUpLine(bool isLeft, int i, int j, Game match)
+        {
+            SetColor(match.Winner, match.FirstTeam.Name);
+
+            int y = GetCursorLeft(isLeft, i);
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: Rename

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+            DrawVerticalLine(isLeft, y, i);
+
+            DrawHorizontalLine(isLeft, y, match.SecondTeam);
+
+            _maxCursorTop = Math.Max(_maxCursorTop, Console.CursorTop);
+        }
+
+        private void DrawHorizontalLine(bool isLeft, int cursorLeft, Team team)
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: Rename method

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+        }
+    }
+
+    public enum GridType
+    {
+        StandardGrid,
+        DoubleGrid
+    }
+
+    public enum LineColor
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: Move to new file

File: 2018/Pichkur/Championship/Championship/Program.cs [hunk 10]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Championship
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            UserInterface a = new UserInterface();
```
Comment (vikian050194) [inline 2018/Pichkur/Championship/Championship/Program.cs]: Put start method here

File: 2018/Pichkur/Championship/Championship/UserInterface.cs [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        {
+            BinaryFormatter formatter = new BinaryFormatter();
+
+            using (FileStream fs = new FileStream("Championship.dat", FileMode.OpenOrCreate))
+            {
+                formatter.Serialize(fs, _data);
+            }
+        }
+
+        public bool Download()
```
Comment (dkblock) [inline 2018/Pichkur/Championship/Championship/UserInterface.cs]: Rename to "load"

File: 2018/Pichkur/Championship/Championship/ChampionshipData.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+
+namespace Championship
+{
+    [Serializable]
+    class ChampionshipData
+    {
+        public SingleEliminationGrid _singleTournamentGrid;
```
Comment (dkblock) [inline 2018/Pichkur/Championship/Championship/ChampionshipData.cs]: Different marking of public variables

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+            return GetWidthOfBranch(i - 1) * 2 + 1;
+        }
+
+        private int GetLengthOfBranch(int i)
+        {
+            return GetWidthOfBranch(i) / 2;
+        }
+    }
+
+    public enum GridType
```
Comment (dkblock) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: May be, move out in new class?

File: 2018/Pichkur/Championship/Championship/UserInterface.cs [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+        {
+            if (_data.isTeamInput)
+                ShowWarning(Messages.TeamsInputAlready);
+            else
+            {
+                _data._doubleTournamentGrid = new DoubleEliminationGrid(DataInput.InputTeams());
+                _data.isTeamInput = true;
+                Save();
+            }
+        }
```
Comment (dkblock) [inline 2018/Pichkur/Championship/Championship/UserInterface.cs]: InputTeamsForSingleTournament() and InputTeamsForDoubleTournament() do the same

File: 2018/Pichkur/Championship/Championship/ChampionshipData.cs (hunk 12 above)
Comment (Ten1n) [reply 2018/Pichkur/Championship/Championship/ChampionshipData.cs]: NamingConvention

File: 2018/Pichkur/Championship/Championship/DoubleEliminationGrid.cs [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            WinnersTour.StartNextTour();
+        }
+
+        private List<Team> GetLoserTeamsFromWinnersTour()
+        {
+            List<Team> teams = new List<Team>();
+
+            foreach (var game in WinnersTour.Tours[WinnersTour.CountTours].Games)
+            {
+                if (game != null)
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/DoubleEliminationGrid.cs]: NullObjectPattern

File: 2018/Pichkur/Championship/Championship/Game.cs [hunk 16]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
+
+namespace Championship
+{
+    [Serializable]
+    public class Game
+    {
+        public Team FirstTeam { get; private set; }
+        public Team SecondTeam { get; private set; }
+        public Team Winner { get; private set; }
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/Game.cs]: Winner vs GetLooser()

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 17]
```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Championship
+{
+    class GridDrawer
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: GridRenderer

File: 2018/Pichkur/Championship/Championship/GridDrawer.cs [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        private void PrintTeamName(bool isleft, int cursorleft, Team team)
+        {
+            if (isleft)
+            {
+                Console.CursorLeft = Math.Max(cursorleft - _shift, 0);
+
+                if (team == null)
+                    Console.WriteLine($"{_defoltTeamName}--");
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/GridDrawer.cs]: Defolt

File: 2018/Pichkur/Championship/Championship/Menu.cs [hunk 19]
```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Championship
+{
+    [Serializable]
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/Menu.cs]: menu should not be serializable

File: 2018/Pichkur/Championship/Championship/MenuItem.cs [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+    [Serializable]
+    public class MenuItem
+    {
+        public MenuItem()
+        {
+            Title = "";
+        }
+        public MenuItem(Action action, string title)
+        {
+            Action += action;
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/MenuItem.cs]: wow

File: 2018/Pichkur/Championship/Championship/Program.cs (hunk 10 above)
Comment (Ten1n) [reply 2018/Pichkur/Championship/Championship/Program.cs]: UI

File: 2018/Pichkur/Championship/Championship/Tour.cs [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+            }
+            else
+            {
+                DistributeExtraTeams(teams);
+            }
+
+            _extraTeams = teams;
+        }
+
+        public Tour(Tour tour)
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/Tour.cs]: rename to previousTour
or pass 2 params instead

File: 2018/Pichkur/Championship/Championship/UserInterface.cs [hunk 22]
```diff
@@ -0,0 +1,264 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Championship
```
Comment (Ten1n) [inline 2018/Pichkur/Championship/Championship/UserInterface.cs]: split logic from UI

**Pull request merged**